service/history/nDCEventsReapplication.go

```diff
@@ -0,0 +1,262 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../LICENSE -package $GOPACKAGE -source $GOFILE -destination newNDCEventsReapplication_mock.go
+
+package history
+
+import (
+	ctx "context"
+	"fmt"
+
+	"github.com/pborman/uuid"
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/clock"
+	"github.com/uber/cadence/common/cluster"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/metrics"
+)
+
+type (
+	nDCEventsReapplication struct {
+		timeSource      clock.TimeSource
+		historyCache    *historyCache
+		domainCache     cache.DomainCache
+		clusterMetadata cluster.Metadata
+		resetor         workflowResetor
+		metricsClient   metrics.Client
+		logger          log.Logger
+	}
+)
+
+func newNDCEventsReapplication(
+	timeSource clock.TimeSource,
+	resetor workflowResetor,
+	historyCache *historyCache,
+	domainCache cache.DomainCache,
+	metricsClient metrics.Client,
+	logger log.Logger,
+	clusterMetadata cluster.Metadata,
+) *nDCEventsReapplication {
+
+	return &nDCEventsReapplication{
+		timeSource:      timeSource,
+		historyCache:    historyCache,
+		domainCache:     domainCache,
+		metricsClient:   metricsClient,
+		logger:          logger,
+		clusterMetadata: clusterMetadata,
+		resetor:         resetor,
+	}
+}
+func (r *nDCEventsReapplication) reapplyEvents(
+	ctx ctx.Context,
+	domainID string,
+	workflowID string,
+	historyEvents []*shared.HistoryEvent,
+) (retError error) {
+	reapplyEvents := []*shared.HistoryEvent{}
```

Comment:
plz add a TODO here, stating that there should be a reapply policy in the future

Reply:
done